GPU/GPUTracking/Merger/GPUTPCGMMerger.cxx

```diff
@@ -1919,3 +1935,136 @@ GPUd() void GPUTPCGMMerger::Finalize2(int nBlocks, int nThreads, int iBlock, int
     }
   }
 }
+
+struct MergeLooperParam {
+  float absz;
+  float tgl;
+  float qpt;
+  float x;
+  float y;
+  unsigned int id;
+};
+
+GPUd() void GPUTPCGMMerger::MergeLoopers(int nBlocks, int nThreads, int iBlock, int iThread)
+{
+  if (iThread || iBlock) {
+    return;
+  }
+#ifndef GPUCA_GPUCODE
+  std::vector<MergeLooperParam> params;
+  const float lowPtThresh = Param().rec.tpcRejectQPt * 1.1f; // Might need to merge tracks above the threshold with parts below the threshold
+  for (unsigned int i = 0; i < mMemory->nOutputTracks; i++) {
+    const auto& trk = mOutputTracks[i];
+    const auto& p = trk.GetParam();
+    const float qptabs = CAMath::Abs(p.GetQPt());
+    if (trk.NClusters() && qptabs > 5.f && qptabs <= lowPtThresh) {
+      const int slice = mClusters[trk.FirstClusterRef() + trk.NClusters() - 1].slice;
+      const float z = p.GetZ() + (Param().earlyTpcTransform ? p.GetTZOffset() : GetConstantMem()->calibObjects.fastTransform->convVertexTimeToZOffset(slice, p.GetTZOffset(), Param().continuousMaxTimeBin));
```

Comment:
If I got it right, you are comparing any pair of tracks which are separated in z by <100 cm, with z defined in the TPC volume (after accounting for time0, otherwise you cannot use this z for field query). But should not one also cut on time0 difference?
Here is the max time0 difference of 2 tracks with the same MC label.
![image](https://user-images.githubusercontent.com/7382029/97874886-f4009980-1d19-11eb-94f8-f6cda1a169cd.png)

For the rest, I am using more or less similar code but trying to find an optimal way to tag >2 legs (the e-loss effect is very high, and having e.g. legs A, B, C, the C will not necessarily match to A but may match to B).

Reply:
I have also seen this energy loss effect, but if I match A to B and B to C, only the one closest to the vertex will survive. It is not necessary to directly match A to C.

From your plot, perhaps 100 cm is actually too tight. On the other hand, I assume that via matching A to B and B to C, I can also cover longer distances in z. The only problem is if 2 adjacent legs are separated by more then 100cm, which is probably not so likely.
And I don't want to allow too large z differences, since the complexity and fake match rate should go with z².

I could probably cut on the time0 difference taking into account the max deltaFwd and deltaBwd to remove impossible matches.

Reply:
To be sure: 
The absz you are saving is the position of the track lowest point within the TPC, right? In this case current code may compare tracks separated by many orbits; to exclude this, I think the time0 should be used.
(Alternatively, if the z is just an alias of L-v*t, then one cannot use if in the field query...)

Reply:
Thx for catching, that was wrong. The field should be fetched via `p.GetZ()`